File: 787-cheapest-flights-within-k-stops/python3/787-cheapest-flights-within-k-stops.py
class Solution:
    def findCheapestPrice(self, n: int, flights: List[List[int]], src: int, dst: int, k: int) -> int:
        edges = [[] for x in range(n)]
        for f,t,c in flights:
            edges[f].append((t, c))
        cost_t = [sys.maxsize] * n
        q = [(src, 0)]
        stops = 0
        while q and stops <= k:
            size = len(q)
            for i in range(size):
                origin, cost = q.pop(0)
            
                for to, price in edges[origin]:
                    cost_to = cost + price
                    if cost_to > cost_t[to]:
                        continue
                    cost_t[to] = cost_to
                    q.append((to, cost_to))
            stops += 1
                    
        return cost_t[dst] if cost_t[dst] != sys.maxsize else -1

File: 787-cheapest-flights-within-k-stops/java/787-cheapest-flights-within-k-stops.java
class Solution {
  public int findCheapestPrice(int n, int[][] flights, int src, int dst, int k) {
    var edges = new HashMap<Integer, List<Pair<Integer, Integer>>>();

    for (var i = 0; i < flights.length; ++i) {
      var from = flights[i][0];
      var to = flights[i][1];
      var price = flights[i][2];

      if (!edges.containsKey(from)) {
        edges.put(from, new ArrayList<>());
      }

      edges.get(from).add(new Pair<>(to, price));
    }

    var cost_dst = new int[n];
    for (var i = 0; i < n; ++i) {
      cost_dst[i] = Integer.MAX_VALUE;
    }

    var queue = new ArrayDeque<Pair<Integer, Integer>>();
    queue.add(new Pair<>(src, 0));
    var stops = 0;

    while (queue.size() > 0 && stops <= k) {
      var size = queue.size();
      for (var i = 0; i < size; ++i) {
        var leg = queue.pollFirst();
        var origin = leg.getKey();
        var cost = leg.getValue();

        for (var t : edges.getOrDefault(origin, List.of())) {
          var destination = t.getKey();
          var price = t.getValue();

          var cost_to = cost + price;
          if (cost_dst[destination] < cost_to) {
            continue;
          }
          cost_dst[destination] = cost_to;
          queue.add(new Pair<>(destination, cost_to));
        }
      }
      ++stops;
    }

    return cost_dst[dst] != Integer.MAX_VALUE ? cost_dst[dst] : -1;
  }
}

